text/0161-pton-standard.md

```diff
@@ -0,0 +1,287 @@
+
+# Guide
+
+## Useful links
+
+1. [Reference pTON implementation](https://github.com/ston-fi/pton-contracts)
+
+# Specification
+
+Proxy TON must be organized as follows: there is minter smart-contract which is used to mint new wallets and provide common information. Total supply must be 0, since minter can't know how many tokenized tokens there are in all wallets.
+
+At the same time information about amount of tokenized TON owned by each user is stores in decentralized manner in individual (for each owner) smart-contracts called "wallets".
+
+## Wallet Contract
+
+Must implement:
+
+### Internal message handlers
+
+#### 1. `ton_transfer`
+
+##### Request
+
+TL-B schema of inbound message:
+
+```
+ton_transfer#1f3835d
+   query_id:uint64 
+   ton_amount:Coins 
+   refund_address:MsgAddress 
+   forward_payload:(Either Cell ^Cell) 
+= InternalMsgBody;
+```
+
+`query_id` - arbitrary request number.
+
+`ton_amount` - amount of transferred TON in elementary units.
+
+`refund_address` - address of the the refund tx, if execution fails.
+
+`forward_payload` - optional custom data that should be sent to the wallet owner address.
+
+**Should be rejected if:**
+
+1. `ton_amount` is zero
+2. `refund_address` is not `addr_std`
+3. message value is less or equal than `ton_amount`
+4. outgoing messages fails
+5. if rejected, should send a `ton_refund` message with following layout TL-B: `ton_refund#ae25d79e query_id:uint64 = InternalMsgBody`
+
+**Otherwise should do:**
+
+1. increase internal jetton balance by `ton_amount`
+2. the receiver's jetton-wallet send message to `owner` address with unused gas attached and with the TEP-74 `transfer_notification` message body
+
+#### 2. `transfer`
+
+##### Request
+
+TL-B schema of inbound message:
+
+```
+transfer#0f8a7ea5 
+   query_id:uint64 
+   amount:(VarUInteger 16) 
+   destination:MsgAddress
+   response_destination:MsgAddress 
+   custom_payload:(Maybe ^Cell)
+   forward_ton_amount:(VarUInteger 16) 
+   forward_payload:(Either Cell ^Cell)
+= InternalMsgBody;
+```
+
+`query_id` - arbitrary request number
+
+`amount` - amount of transferred TON in elementary units
+
+`destination` - recipient that will receive TON
+
+`response_destination` - ignored
+
+`custom_payload` - ignored
+
+`forward_ton_amount` - ignored
+
+`forward_payload` - optional custom data that should be sent to the destination address
+
+**Should be rejected if:**
+
+1. message is not from the owner.
+2. there is no enough tokenized ton on the sender wallet
+3. there is no enough TON (with respect to jetton own storage fee guidelines and operation costs) to process operation
+
+**Otherwise should do:**
+
+1. decrease ton balance on sender wallet by `amount`
+2. send a message to receiver with the `ton_transfer` layout described above
```

Comment:
`ton_amount` is the amount of ton to be tokenized, this amount is sent in `transfer_notification`, the rest of the ton is used as gas (of course tx value must be greater than specified `ton_amount` or it will throw)